### PR TITLE
MueLu: region w-cycle and symmetric Gauss-Seidel

### PR DESCRIPTION
**Pull request by pohm01**
@trilinos/muelu 

## Motivation
Adds the options for a W-cycle and Symmetric Gauss-Seidel smoother for region multigrid.

### CODE REVIEW

File: packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionSmoothers_def.hpp [hunk 1]
```diff
@@ -260,6 +262,25 @@ void GSIterate(RCP<Teuchos::ParameterList> smootherParams,
       OneregX[k] += ldelta[k];
     }
     zeroInitGuess = false;
+
+    if( sgs ){
+      for (size_t k = numRows; k--;) {
```
Comment (mayrmt) [inline packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionSmoothers_def.hpp:267]: Shouldn't this be
```suggestion
      for (size_t k = numRows - 1; k >= 0; k--) {
```
to start looping at the last row and then move backward to the first row?

Comment (pohm01) [reply packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionSmoothers_def.hpp:267]: I'm not sure why, but `for (size_t k = numRows - 1; k >= 0; k--)` does not work. It is some sort of c++ thing that I don't fully understand.
I did check that `for (size_t k = numRows; k--;)` correctly iterates from last row to first row (starts at `numRows-1` and ends at `0`).

Comment (mayrmt) [reply packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionSmoothers_def.hpp:267]: I was just curious. If it works, then I'm fine with it.

Comment (brian-kelley) [reply packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionSmoothers_def.hpp:267]: @pohm01 @mayrmt size_t is unsigned, so if it's 0 and you decrement it, it wraps around to 2^64 - 1. So k >= 0 is true no matter what k is. You could use signed counter, or loop k from 0 to numRows and then use ``numRows - k - 1`` in the loop.

**Comment by mayrmt** [comment]
@pohm01 Feel free to merge this, whenever you think this is ready.

**Pull request merged**